ddtrace/contrib/cassandra/session.py

```diff
@@ -31,32 +36,133 @@ def traced_connect(func, instance, args, kwargs):
     return session
 
-def traced_execute(func, instance, args, kwargs):
-    cluster = getattr(instance, 'cluster', None)
+def _close_span_on_success(result, future):
+    span = getattr(future, CURRENT_SPAN, None)
+    if not span:
+        log.debug('traced_set_final_result was not able to get the current span from the ResponseFuture')
+        return
+    with span:
+        span.set_tags(_extract_result_metas(cassandra.cluster.ResultSet(future, result)))
+
+def traced_set_final_result(func, instance, args, kwargs):
+    result = kwargs.get('result') or args[0]
+    _close_span_on_success(result, instance)
+    return func(*args, **kwargs)
+
+def _close_span_on_error(exc, future):
+    span = getattr(future, CURRENT_SPAN, None)
+    if not span:
+        log.debug('traced_set_final_exception was not able to get the current span from the ResponseFuture')
+        return
+    with span:
+        # FIXME how should we handle an exception that hasn't be rethrown yet
+        try:
+            raise exc
+        except:
+            span.set_exc_info(*sys.exc_info())
+
+def traced_set_final_exception(func, instance, args, kwargs):
+    exc = kwargs.get('result') or args[0]
+    _close_span_on_error(exc, instance)
+    return func(*args, **kwargs)
+
+def traced_start_fetching_next_page(func, instance, args, kwargs):
+    has_more_pages = instance._paging_state is not None
+    if not has_more_pages:
+        return func(*args, **kwargs)
+    session = getattr(instance, 'session', None)
+    cluster = getattr(session, 'cluster', None)
     pin = Pin.get_from(cluster)
     if not pin or not pin.enabled():
         return func(*args, **kwargs)
 
-    service = pin.service
-    tracer = pin.tracer
+    # In case the current span is not finished we make sure to finish it
+    old_span = getattr(instance, CURRENT_SPAN, None)
+    if old_span:
+        log.debug('previous span was not finished before fetching next page')
+        old_span.finish()
 
-    query = kwargs.get("kwargs") or args[0]
+    query = getattr(instance, 'query', None)
+    page_number = getattr(instance, PAGE_NUMBER, 0) + 1
 
-    with tracer.trace("cassandra.query", service=service, span_type=cassx.TYPE) as span:
-        _sanitize_query(span, query)
-        span.set_tags(_extract_session_metas(instance))     # FIXME[matt] do once?
-        span.set_tags(_extract_cluster_metas(cluster))
-        result = None
-        try:
-            result = func(*args, **kwargs)
-            return result
-        finally:
-            if result:
-                span.set_tags(_extract_result_metas(result))
+    span = _start_span_and_set_tags(pin, query, session, cluster, page_number)
+
+    setattr(instance, PAGE_NUMBER, page_number)
+    setattr(instance, CURRENT_SPAN, span)
+    try:
+        return func(*args, **kwargs)
+    except:
+        span.set_exc_info(*sys.exc_info())
+        span.finish()
```

Comment:
put the `span.finish()` in the `finally` block so it's closed even if we have an exception in `set_exc_info()`

Reply:
Putting `span.finish()` in a `finally` would close the span even when the method returns without exception, which is not the intended behavior. I'll put a `with` block in the `except` block instead.

Reply:
ok so if we don't have exceptions it must not be closed at this stage

Reply:
That's correct.